### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [1]:
# Dependencies and Setup
import pandas as pd

In [2]:
# File to Load
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas dataframe
purchase_data = pd.read_csv(file_to_load)

## Player Count

* Display the total number of players


In [3]:
player_count = len(purchase_data['SN'].unique())

player_count_df = pd.DataFrame([{"Total Players" : player_count}])
player_count_df

Total Players
0            576

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [4]:
items_unique = len(purchase_data['Item ID'].unique())

item_avg_price = purchase_data['Price'].mean()

item_purchases = len(purchase_data['Purchase ID'])

item_revenue = item_avg_price * item_purchases

purchasing_analysis_df = pd.DataFrame([{
    "Number of Unique Items" : items_unique,
    "Average Price" : round(item_avg_price, 2),
    "Number of Purchases" : item_purchases,
    "Total Revenue" : item_revenue
}])

purchasing_analysis_df["Average Price"] = purchasing_analysis_df["Average Price"].map("${:,.2f}".format)
purchasing_analysis_df["Total Revenue"] = purchasing_analysis_df["Total Revenue"].map('${:,.2f}'.format)

purchasing_analysis_df

Number of Unique Items Average Price  Number of Purchases Total Revenue
0                     179         $3.05                  780     $2,379.77

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [5]:
gender_sn_unique = purchase_data.drop_duplicates(['SN', 'Gender'])

gender_sn_count = gender_sn_unique['Gender'].value_counts()
gender_sn_count_df = pd.DataFrame(gender_sn_count)

gender_sn_percent = gender_sn_unique['Gender'].value_counts('SN')
gender_sn_percent_df = pd.DataFrame(gender_sn_percent)

gender_sn_count_df['Percentage of Players'] = gender_sn_percent_df['Gender']

gender_demographics_df = gender_sn_count_df.rename(columns = {'Gender' : 'Total Count'})

gender_demographics_df["Percentage of Players"] = gender_demographics_df["Percentage of Players"].map("{:.2%}".format)

gender_demographics_df

Total Count Percentage of Players
Male                           484                84.03%
Female                          81                14.06%
Other / Non-Disclosed           11                 1.91%

## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [6]:
gender_groupby = purchase_data.groupby("Gender")
gender_person_groupby = purchase_data.groupby(["SN", "Gender"])

gender_purchasecount = gender_groupby["Purchase ID"].count()
gender_purchasecount_df = pd.DataFrame(gender_purchasecount)

gender_avgprice = gender_groupby["Price"].mean()
gender_avgprice_df = pd.DataFrame(gender_avgprice)

gender_totalprice = gender_groupby["Price"].sum()
gender_totalprice_df = pd.DataFrame(gender_totalprice)

gender_person_avgprice_total = gender_person_groupby["Price"].sum()
gender_person_avgprice_total_df = pd.DataFrame(gender_person_avgprice_total) 

new_gender_person_groupby = gender_person_avgprice_total_df.groupby("Gender").mean()
new_gender_person_groupby_df = pd.DataFrame(new_gender_person_groupby)

purchasing_analysis_gender_df = gender_purchasecount_df
purchasing_analysis_gender_df = purchasing_analysis_gender_df.rename(columns = {'Purchase ID' : 'Purchase Count'})
purchasing_analysis_gender_df["Average Purchase Price"] = round(gender_avgprice_df, 2)
purchasing_analysis_gender_df["Total Purchase Value"] = gender_totalprice_df
purchasing_analysis_gender_df["Avg Total Purchase per Person"] = round(new_gender_person_groupby_df, 2)

purchasing_analysis_gender_df["Average Purchase Price"] = purchasing_analysis_gender_df["Average Purchase Price"].map("${:,.2f}".format)
purchasing_analysis_gender_df["Total Purchase Value"] = purchasing_analysis_gender_df["Total Purchase Value"].map("${:,.2f}".format)
purchasing_analysis_gender_df["Avg Total Purchase per Person"] = purchasing_analysis_gender_df["Avg Total Purchase per Person"].map("${:,.2f}".format)

purchasing_analysis_gender_df

Purchase Count Average Purchase Price  \
Gender                                                         
Female                            113                  $3.20   
Male                              652                  $3.02   
Other / Non-Disclosed              15                  $3.35   

                      Total Purchase Value Avg Total Purchase per Person  
Gender                                                                    
Female                             $361.94                         $4.47  
Male                             $1,967.64                         $4.07  
Other / Non-Disclosed               $50.19                         $4.56

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()



* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [7]:
# Binning
age_list = ["<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+"]
age_bins = [0, 10, 15, 20, 25, 30, 35, 40, 1000]

age_data = gender_sn_unique

age_data["Age Ranges"] = pd.cut(age_data["Age"], bins = age_bins, labels = age_list, right = False, include_lowest = True)


# Grouping

# age_groupby = age_data.groupby("Age Group")
# age_count = pd.DataFrame(age_groupby["SN"].count())
# age_count

age_count = age_data['Age Ranges'].value_counts()
age_count_df = pd.DataFrame(age_count)

age_percent = age_data['Age Ranges'].value_counts('SN')
age_percent_df = pd.DataFrame(age_percent)

age_count_df['Percentage of Players'] = round(age_percent_df['Age Ranges'], 4)

age_demographics_df = age_count_df.rename(columns = {'Age Ranges' : 'Total Count'})
age_demographics_df.sort_index(inplace = True)

age_demographics_df["Percentage of Players"] = age_demographics_df["Percentage of Players"].map("{:.2%}".format)

age_demographics_df

<ipython-input-7-84e4a1d22d45>:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  age_data["Age Ranges"] = pd.cut(age_data["Age"], bins = age_bins, labels = age_list, right = False, include_lowest = True)


Total Count Percentage of Players
<10             17                 2.95%
10-14           22                 3.82%
15-19          107                18.58%
20-24          258                44.79%
25-29           77                13.37%
30-34           52                 9.03%
35-39           31                 5.38%
40+             12                 2.08%

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [8]:
age_data_full = purchase_data
age_data_full["Age Ranges"] = pd.cut(age_data_full["Age"], bins = age_bins, labels = age_list, right = False, include_lowest = True)

age_groupby = age_data_full.groupby("Age Ranges")
age_person_groupby = age_data_full.groupby(["SN", "Age Ranges"])

age_purchasecount = age_groupby["Purchase ID"].count()
age_purchasecount_df = pd.DataFrame(age_purchasecount)

age_avgprice = age_groupby["Price"].mean()
age_avgprice_df = pd.DataFrame(age_avgprice)

age_totalprice = age_groupby["Price"].sum()
age_totalprice_df = pd.DataFrame(age_totalprice)

age_person_avgprice_total = age_person_groupby["Price"].sum()
age_person_avgprice_total_df = pd.DataFrame(age_person_avgprice_total)

new_age_person_groupby = age_person_avgprice_total_df.groupby("Age Ranges").mean()
new_age_person_groupby_df = pd.DataFrame(new_age_person_groupby)

purchasing_analysis_age_df = age_purchasecount_df
purchasing_analysis_age_df = purchasing_analysis_age_df.rename(columns = {'Purchase ID' : 'Purchase Count'})
purchasing_analysis_age_df["Average Purchase Price"] = round(age_avgprice_df, 2)
purchasing_analysis_age_df["Total Purchase Value"] = age_totalprice_df
purchasing_analysis_age_df["Avg Total Purchase per Person"] = round(new_age_person_groupby_df, 2)

purchasing_analysis_age_df["Average Purchase Price"] = purchasing_analysis_age_df["Average Purchase Price"].map("${:,.2f}".format)
purchasing_analysis_age_df["Total Purchase Value"] = purchasing_analysis_age_df["Total Purchase Value"].map("${:,.2f}".format)
purchasing_analysis_age_df["Avg Total Purchase per Person"] = purchasing_analysis_age_df["Avg Total Purchase per Person"].map("${:,.2f}".format)

purchasing_analysis_age_df

Purchase Count Average Purchase Price Total Purchase Value  \
Age Ranges                                                               
<10                     23                  $3.35               $77.13   
10-14                   28                  $2.96               $82.78   
15-19                  136                  $3.04              $412.89   
20-24                  365                  $3.05            $1,114.06   
25-29                  101                  $2.90              $293.00   
30-34                   73                  $2.93              $214.00   
35-39                   41                  $3.60              $147.67   
40+                     13                  $2.94               $38.24   

           Avg Total Purchase per Person  
Age Ranges                                
<10                                $4.54  
10-14                              $3.76  
15-19                              $3.86  
20-24                              $4.32  
25-29                              $3.81  
30-34                              $4.12  
35-39                              $4.76  
40+                                $3.19

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [9]:
sn_groupby = purchase_data.groupby("SN")

sn_count = sn_groupby["SN"].count()
sn_count_df = pd.DataFrame(sn_count)

sn_avgprice = sn_groupby["Price"].mean()
sn_avgprice_df = pd.DataFrame(sn_avgprice)

sn_total = sn_groupby["Price"].sum()
sn_total_df = pd.DataFrame(sn_total)

top_spenders_df = sn_count_df
top_spenders_df = top_spenders_df.rename(columns = {'SN' : 'Purchase Count'})
top_spenders_df["Average Purchase Price"] = round(sn_avgprice_df, 2)
top_spenders_df["Total Purchase Value"] = sn_total_df

top_spenders_df = top_spenders_df.nlargest(5, "Total Purchase Value")

top_spenders_df["Average Purchase Price"] = top_spenders_df["Average Purchase Price"].map("${:,.2f}".format)
top_spenders_df["Total Purchase Value"] = top_spenders_df["Total Purchase Value"].map("${:,.2f}".format)

top_spenders_df

Purchase Count Average Purchase Price Total Purchase Value
SN                                                                     
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, average item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [10]:
pop_item_groupby = purchase_data.groupby(["Item ID", "Item Name"])

pop_item_count = pop_item_groupby["Item ID"].count()
pop_item_count_df = pd.DataFrame(pop_item_count)

pop_item_price = pop_item_groupby["Price"].mean()
pop_item_price_df = pd.DataFrame(pop_item_price)

pop_item_total = pop_item_groupby["Price"].sum()
pop_item_total_df = pd.DataFrame(pop_item_total)

popular_items_df = pop_item_count_df
popular_items_df = popular_items_df.rename(columns = {'Item ID' : 'Purchase Count'})
popular_items_df["Item Price"] = round(pop_item_price_df, 2)
popular_items_df["Total Purchase Value"] = pop_item_total_df

profitable_items_df = popular_items_df

popular_items_df = popular_items_df.sort_values(by = ["Purchase Count"], ascending = False)

popular_items_df = popular_items_df.nlargest(5, "Purchase Count")

popular_items_df["Item Price"] = popular_items_df["Item Price"].map("${:,.2f}".format)
popular_items_df["Total Purchase Value"] = popular_items_df["Total Purchase Value"].map("${:,.2f}".format)

popular_items_df

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
132     Persuasion                                                 9   
108     Extraction, Quickblade Of Trembling Hands                  9   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
145     Fiery Glass Crusader                              $4.58   
132     Persuasion                                        $3.22   
108     Extraction, Quickblade Of Trembling Hands         $3.53   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
132     Persuasion                                                 $28.99  
108     Extraction, Quickblade Of Trembling Hands                  $31.77

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [11]:
profitable_items_df = profitable_items_df.nlargest(5, "Total Purchase Value")

profitable_items_df["Item Price"] = profitable_items_df["Item Price"].map("${:,.2f}".format)
profitable_items_df["Total Purchase Value"] = profitable_items_df["Total Purchase Value"].map("${:,.2f}".format)

profitable_items_df

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
103     Singed Scalpel                                             8   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
103     Singed Scalpel                                    $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
103     Singed Scalpel                                             $34.80

## File Outputs
- Output all Pandas dataframes as .csv files
- These should be outputted to the 'Outputs' folder in root

In [12]:
player_count_df.to_csv("../Outputs/player_count.csv")
purchasing_analysis_df.to_csv("../Outputs/purchasing_analysis.csv")
gender_demographics_df.to_csv("../Outputs/gender_demographics.csv")
purchasing_analysis_gender_df.to_csv("../Outputs/purchasing_analysis.csv")
age_demographics_df.to_csv("../Outputs/age_demographics.csv")
purchasing_analysis_age_df.to_csv("../Outputs/purchasing_analysis_age.csv")
top_spenders_df.to_csv("../Outputs/top_spenders.csv")
popular_items_df.to_csv("../Outputs/popular_items.csv")
profitable_items_df.to_csv("../Outputs/profitable_items_df.csv")

## Conclusions
- The focus on audience should largely be on 20 to 24 year old players as they not only make up ~45% of the player base, they account for a substantial proportion of spending ($1,114.06).
- While there are a number of users with consistent spending habits, there exist no individual 'whale' users.
- 'The Final Critic' and 'Oathbreaker, Last Hope of the Breaking Storm' should be pushed more heavily as they are conisistent buys with high purchase totals. This may be reflective of their in-game niche as affordable and useful items.